src/python/pants/backend/docker/dependencies.py

```diff
@@ -0,0 +1,40 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+
+from pants.backend.docker.parser import DockerfileInfo, DockerfileParseRequest
+from pants.backend.docker.target_types import DockerDependencies, DockerImageSources
+from pants.engine.addresses import Address, Addresses, UnparsedAddressInputs
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import InjectDependenciesRequest, InjectedDependencies, WrappedTarget
+from pants.engine.unions import UnionRule
+
+
+class InjectDockerDependencies(InjectDependenciesRequest):
+    inject_for = DockerDependencies
+
+
+@rule
+async def inject_docker_dependencies(request: InjectDockerDependencies) -> InjectedDependencies:
+    """Inspects COPY instructions in the Dockerfile for references to known targets."""
+    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
+    sources = original_tgt.target[DockerImageSources]
+    if not sources.value:
+        return InjectedDependencies()
+
+    dockerfile = await Get(DockerfileInfo, DockerfileParseRequest(sources))
+    addresses = await Get(
+        Addresses,
+        UnparsedAddressInputs(
+            dockerfile.putative_target_addresses,
+            owning_address=original_tgt.target.address,
+        ),
+    )
```

Comment:
Do we want to check that each target does in fact have a `PexBinaryFieldSet`? And we could even then check that the path to the pex it generates is the one we originally found in the COPY statement? Might be overkill though.

Reply:
That it is a target of expected type is probably good enough. Been thinking that perhaps we want to expand on what we support to infer to, such as `files` targets or `python_distribution` ?